Objects/unionobject.c

```diff
@@ -137,102 +137,159 @@ union_richcompare(PyObject *a, PyObject *b, int op)
     return result;
 }
 
+static int
+is_same(PyObject* left, PyObject* right)
+{
+    int is_ga = _PyGenericAlias_Check(left) && _PyGenericAlias_Check(right);
+    return is_ga ? PyObject_RichCompareBool(left, right, Py_EQ) : left == right;
+}
+
+static int
+is_unionable(PyObject *obj)
+{
+    return (obj == Py_None ||
+        PyType_Check(obj) ||
+        _PyGenericAlias_Check(obj) ||
+        _PyUnion_Check(obj));
+}
+
 static PyObject*
-flatten_args(PyObject* args)
+merge_union_and_union(PyObject *left, PyObject *right)
 {
-    Py_ssize_t arg_length = PyTuple_GET_SIZE(args);
-    Py_ssize_t total_args = 0;
-    // Get number of total args once it's flattened.
-    for (Py_ssize_t i = 0; i < arg_length; i++) {
-        PyObject *arg = PyTuple_GET_ITEM(args, i);
-        if (_PyUnion_Check(arg)) {
-            total_args += PyTuple_GET_SIZE(((unionobject*) arg)->args);
-        } else {
-            total_args++;
-        }
-    }
-    // Create new tuple of flattened args.
-    PyObject *flattened_args = PyTuple_New(total_args);
-    if (flattened_args == NULL) {
+    PyObject* left_args = ((unionobject *) left)->args;
+    PyObject* right_args = ((unionobject*) right)->args;
+
+    Py_ssize_t args_length = PyTuple_GET_SIZE(left_args);
+    Py_ssize_t other_args_length = PyTuple_GET_SIZE(right_args);
+
+    PyObject *tuple = PyTuple_New(args_length + other_args_length);
+
+    if (tuple == NULL) {
         return NULL;
     }
-    Py_ssize_t pos = 0;
-    for (Py_ssize_t i = 0; i < arg_length; i++) {
-        PyObject *arg = PyTuple_GET_ITEM(args, i);
-        if (_PyUnion_Check(arg)) {
-            PyObject* nested_args = ((unionobject*)arg)->args;
-            Py_ssize_t nested_arg_length = PyTuple_GET_SIZE(nested_args);
-            for (Py_ssize_t j = 0; j < nested_arg_length; j++) {
-                PyObject* nested_arg = PyTuple_GET_ITEM(nested_args, j);
-                Py_INCREF(nested_arg);
-                PyTuple_SET_ITEM(flattened_args, pos, nested_arg);
-                pos++;
+
+    for (int i = 0; i < args_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(left_args, i);
+        Py_INCREF(arg);
+        PyTuple_SET_ITEM(tuple, i, arg);
+    }
+
+    Py_ssize_t pos = args_length;
+
+    for (int i = 0; i < other_args_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(right_args, i);
+
+        int is_duplicate = 0;
+
+        for (int j = 0; j < args_length; j++) {
+            PyObject *left_arg = PyTuple_GET_ITEM(left_args, j);
+            is_duplicate = is_same(left_arg, arg);
+
+            if (is_duplicate < 0) {
+                Py_DECREF(tuple);
+                return NULL;
             }
-        } else {
-            if (arg == Py_None) {
-                arg = (PyObject *)&_PyNone_Type;
+            if (is_duplicate) {
+                break;
             }
+        }
+
```

Comment:
Extract the code which checks whether the object is in the tuple into a separate function or just use `PySequence_Contains()` (if it does not add significant overhead).

Reply:
Done